codePost_api/helpers.py

```diff
@@ -184,7 +189,45 @@ class UploadModes(_DocEnum):
 # Core upload logic
 ###########################################################################################
 class UploadError(RuntimeError):
-    pass
+
+    def __init__(self, message, api_key=None, assignment_id=None, submission_id=None, file_ids=None, *args, **kwargs):
+        super(UploadError, self).__init__(message, *args, **kwargs)
+        
+        self.api_key = api_key
+        self._assignment_id = assignment_id
+        
+        # Initialize object lists
+        self._submission_ids = [] if submission_id == None else [ submission_id ]
+        self._file_ids = [] if file_ids == None else file_ids
+
+        # Trace of operations undertaken by the cleanup method
+        self._trace = ""
+    
+    def force_cleanup(self):
+
+        self._trace += "<<"
+        
+        for f_id in self._file_ids:
+            try:
+                delete_file(api_key=self.api_key, file_id=f_id)
+            except:
+                self._trace += "!file({}); ".format(f_id)
+            else:
+                self._trace += "-file({}); ".format(f_id)
+        
+        for s_id in self._submission_ids:
+            try:
```

Comment:
Is `_submission_ids` ever going to contain more than 1 id?

Reply:
I was leaving the door open for this exception to be used in other scenarios. For instance, to report a collision between existing submissions, as is the case in L287-299 of that file:
https://github.com/codepost-io/codePost-api-python/blob/ef24aeee9119786fcb9e5ab79df058b466a4e02d/codePost_api/helpers.py#L287-L299
Currently these errors are reported with an error message, but it would be cleaner to use this structure (an exception with additional metadata) everywhere: First, this avoids having these long templated exception message strings everywhere; second, it will likely allow us to be smarter at handling exceptions, in the same vein as here where the metadata allows us to know what to clean-up.

Reply:
@jamesaevans Thoughts about this?

Reply:
I agree with you. Good call.